File: 01_Easy/13. Roman to Integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        Roman_dict = {"I":1, "V": 5,"X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
        Roman2_dict = {"IV": 4, "IX": 9, "XL": 40, "XC": 90, "CD": 400, "CM": 900}
        ans = 0
        for key in Roman2_dict:
            if key in s:
                ans += Roman2_dict[key]
                s = s.replace(key, "", 1)
        
        for c in s:
            ans += Roman_dict[c]
        
        return ans
        
if __name__ == '__main__':
    s = "MCMXCIV"
    sol = Solution()
    print(sol.romanToInt(s))

File: CPP/hashmap/13. Roman to Integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        unordered_map<char, int> roman_int_map {
            {'I', 1}, {'V', 5}, {'X', 10}, {'L', 50},
            {'C', 100}, {'D', 500}, {'M', 1000}
        };

        int res = 0;
        for (int i = 0; i < s.size(); i++){
            if (roman_int_map[s[i]] < roman_int_map[s[i+1]]){
                res += roman_int_map[s[i+1]] - roman_int_map[s[i]];
                i++;
            }
            else {
                res += roman_int_map[s[i]];
            }
        }
        return res;
    }
};